zulip_bots/zulip_bots/bots/virtual_fs/virtual_fs.py

```diff
@@ -18,16 +30,29 @@ def handle_message(self, message, bot_handler, state_handler):
         if isinstance(recipient, list):  # If not a stream, then hash on list of emails
             recipient = " ".join([x['email'] for x in recipient])
 
-        with state_handler.state({}) as state:
-            if recipient not in state:
-                state[recipient] = fs_new()
-            fs = state[recipient]
+        # If state_handler is None, we are an embedded bot. Then, state is a
+        # dictionary-like database table and can be accessed via bot_handler.state.
+        if state_handler is None:
+            if recipient not in bot_handler.state:
+                bot_handler.state[recipient] = json.dumps(fs_new(), cls=SetEncoder)
+            fs = json.loads(bot_handler.state[recipient], object_hook=json_list_as_python_set)
             if sender not in fs['user_paths']:
                 fs['user_paths'][sender] = '/'
             fs, msg = fs_command(fs, sender, command)
             prependix = '{}:\n'.format(sender)
             msg = prependix + msg
-            state[recipient] = fs
+            bot_handler.state[recipient] = json.dumps(fs, cls=SetEncoder)
+        else:
+            with state_handler.state({}) as state:
+                if recipient not in state:
+                    state[recipient] = fs_new()
+                fs = state[recipient]
+                if sender not in fs['user_paths']:
+                    fs['user_paths'][sender] = '/'
+                fs, msg = fs_command(fs, sender, command)
+                prependix = '{}:\n'.format(sender)
+                msg = prependix + msg
+                state[recipient] = fs
```

Comment:
This feels like it has a lot of duplicated code flow.  I think what we want is to modify the `state_handler.state` function to accept as an argument the marshalling function for each direction and just have the library take care of the cases; then we can still use the nice `with` logic.

Reply:
It does feel funny to me that virtual_fs has to care whether it's in an embedded bot or not.  I would like there's always a StateHandler, and the implementations are different on the backend.

Reply:
I'll shortly push an update which reduces the boilerplate here to choosing between `bot_handler.StateHandler` and `state_handler.StateHandler`.  I see this remaining check as unavoidable as long as we don't upgrade the external bot's state handler to use the server state handler.

Reply:
I've been out of the loop, because I didn't realize we had two state handlers.  I was under the mindset that you'd only ever have one API to write state, and then the deployment option would govern how state got saved on the backend, plus whether state persisted if you restart the bot.

Reply:
We could've gone the way of contuing to use the old statehandler and cranking it so it fits the new format, but I felt that the conceptual differences were too big: The old StateHandler was an object that could be assigned any value, so of type Any, the new StateHandler behaves like a Dict[str, str]. Furthermore, the old statehandler got passed to the bot in handle_message, whereas the new one is an attribute of the bot_handler. I should've probably discussed that change with someone, it felt cleaner to me. Because of these differences, I thought it'd make sense to have a clear cut between both state handlers. eventually, the old one would be rewritten to fit the new model.

Reply:
I think we might as well eliminate the old one now, since basically nobody uses the old one now, except for the file that you're modifying.

Reply:
As in removing it and making incrementor and virtual_fs embedded_bot-only bots?

Reply:
No, I'm suggesting that we just have them use a StateHandler that works locally.  I'm still a little confused as to why this is difficult.  The whole reason the original code passed in a state_handler to `handle_message` was to give us the flexibility to have the **runtime** control the implementation, not the virtual_fs bot.  It's a slightly leaky abstraction in cases where you don't have permanent file storage, but the original bot had reasonable persistence.

Most bots deal with so little data--even the tricked out virtual_fs bot--that it's totally reasonable to just save a blob after every message.

And my thought was we were simply making the storage mechanisms more permanent here.

Also, there's no reason why a runtime for a non-embedded bot can't write to the db.  We could totally create a save-user-data endpoint.  That was what I thought we were going to do here, but I obviously either didn't communicate it well, or maybe I just don't understand the tradeoffs well enough.